File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head) return NULL;
        unordered_map<Node*,Node*> old;
        old[NULL]=NULL;
        // copy the nodes
        Node* cur=head;
        while(cur){            
            old[cur]=new Node(cur->val);;
            cur=cur->next;
        }
        // link the links
        cur=head;
        while(cur){            
            Node *tmp=old[cur];
            tmp->next = old[cur->next];
            tmp->random = old[cur->random];
            cur=cur->next;
        }
        return old[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        cur=head
        old={None:None}

        while cur:
            tmp=Node(cur.val)
            old[cur]=tmp
            cur=cur.next
        
        cur=head
        while cur:
            tmp=old[cur]
            tmp.next = old[cur.next]
            tmp.random = old[cur.random]
            cur=cur.next
        
        return old[head]